lib/libc/hash/hmac.3
.\" $NetBSD: hmac.3,v 1.5.14.1 2018/06/25 07:25:34 pgoyette Exp $
.\"
.\" Copyright (c) 2016 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Christos Zoulas.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\"TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd May 22, 2018
.Dt HMAC 3
.Os
.Sh NAME
.Nm hmac
.Nd compute a key-Hash Message Authentication Code
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In stdlib.h
.Ft ssize_t
.Fn hmac "const char *hname" "const void *key" "size_t klen" "const void *text" "size_t tlen" "void *digest" "size_t dlen"
.Sh DESCRIPTION
The
.Fn hmac
function computes the key-Hash Message Authentication Code per
.Tn RFC 2104
and places the result in
.Fa digest
writing up to
.Fa dlen
bytes.
The actual number of bytes that would be written is returned.
.Pp
The hash functions supported are: md2, md4, md5, rmd160, sha1, sha224,
sha256, sha384, and sha512.
.Sh RETURN VALUES
The
.Fn hmac
function returns
.Dv \-1
if the
.Fa hname
is not found.
Otherwise the actual length of the digest string is returned (which could
be bigger or smaller than
.Fa dlen ) .
This length depends on the hashing function selected.
.Sh SEE ALSO
.Xr md2 3 ,
.Xr md4 3 ,
.Xr md5 3 ,
.Xr openssl_HMAC 3 ,
.Xr openssl_MD2 3 ,
.Xr openssl_MD4 3 ,
.Xr openssl_MD5 3 ,
.Xr rmd160 3 ,
.Xr sha1 3 ,
.Xr sha2 3
.Sh STANDARDS
.Rs
.%R RFC 2104
.Re
.Sh NOTES
The maximum digest length has been extended from
.Dv 64
to
.Dv 128
bytes to handle SHA2.
.Sh HISTORY
The
.Fn hmac
function appeared in
.Nx 8 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2405 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 8.0 to NetBSD 9.0:
 	byacc: Update to 20170430. [christos 20170605]
 	vax: Add support for VAXstation 4000 TURBOchannel. [flxd 20170609]
 	wsbell(4): Added console bell support for all speaker devices, not
		only those attached at pcppi. [nat 20170612]
 	bind: Import version 9.10.5-P1. [christos 20170615]
	dts: Import dts files from Linux 4.11.5. [jmcneill 20170615]
	amd64: Support 16TB of PA, and 32TB of kernel VA on the architecture
		side. [maxv 20170617]
	expat: Import 2.2.1 (security fixes). [christos 20170617]
	arm: Add support for Allwinner H3 SoC. [jmcneill 20170628]
	bind: Import version 9.10.5-P2. [christos 20170630]
	sunxiemac(4): Add support for Allwinner Gigabit Ethernet (EMAC).
		[jmcneill 20170701]
	wbsio(4): Add support for Winbond W83627DHG-P, W83627SF, W83627UHG,
		W83667HGB, W83687THF, W83697UG,	Nuvoton NCT5104D, NCT6775,
		NCT6779, NCT6791, NCT6792 and NCT6793. [msaitoh 20170707]
	arm: Add support for Allwinner A83T SoC. [jmcneill 20170707]
	MAKEDEV(8): Regenerate, documenting lua, dk, plcom, wmcom, hdmicec, mfi,
		nvme, nvme ns, tpm, and dtrace devices. [sevan 20170710]
	wbsio(4): Add support for NCT6795. [msaitoh 20170711]
	lm(4): Add support for NCT5174D, NCT6775F, NCT6779D and NCT679[1235]D.
		[msaitoh 20170711]
	arm: Add support for NVIDIA Tegra X1 SoC. [jmcneill 20170720]
	itesio(4): Add support IT8628E. [hauke 20170814]
	ichsmb(4): Add support for Intel C3000. [msaitoh 20170817]
	ismt(4): Add support for Intel C3000. [msaitoh 20170817]
	powerpc: Bump MAXTSIZ to 128MB for OEA based systems. [sevan 20170818]
	GMP/MPFR/MPC: Update to 6.1.2, 3.1.5 and 1.0.3. [mrg 20170821]
	ixg(4), ixv(4): Add C3000 support. Add bypass function support for
		bypass adapters. [msaitoh 20170830]
	file(1): Upgraded to 5.32. [christos 20170908]
	tcpdump(8): Import 4.9.2. [christos 20170908]
	veriexecgen(8): Drop support for MD5, SHA1, RMD160. [sevan 20170909]
	itesio(4): Add support IT8728GF and IT877[12]E [hauke 20170814]
	veriexec(4): Drop support for VERIFIED_EXEC_FP_MD5,
		VERIFIED_EXEC_FP_SHA1, and VERIFIED_EXEC_FP_RMD160 options.
		[sevan 20170913]
	acpi(4): Updated ACPICA to 20170831. [christos 20170915]
	dtrace: Install dtruss, execsnoop, opensnoop, and procsystime from the
		DTrace Toolkit if DTrace support is enabled. [sevan 20170917]
	dhcpcd(8): Import dhcpcd-7.0.0-rc3. [roy 20171007]
	ahcisata(4), siisata(4): Add support for NCQ. [jdolecek 20171007]
	mvsata(4): Add support for NCQ, and enable ATAPI support.
		[jdolecek 20171007]
	OpenSSH: Imported 7.6. [christos 20171007]
	amd64: Add support for Kernel ASLR [maxv 20171010]
	tmux(1): Imported 2.6. [christos 20171012]
	amd64: Add support for SMAP. [maxv 20171017]
	amd64: Add support for USER_LDT. [maxv 20171021]
	tzdata updated to 2017c [kre 20171024]
	libc: Update to tzcode2017c. [christos 20171024]
	bwfm(4): Add driver for Broadcom BCM43xxx "FullMAC" USB wireless
		adapters. [jmcneill 20171025]
	kernhist(9): Reworked the export-via-sysctl code to use uintptr_t
		arguments, PR kern/52639.  Requires contemporaneous update
		of vmstat(1). [pgoyette 20171028]
	acpi(4): Updated ACPICA to 20171110. [christos 20171111]
	sunxinand(4): Add driver for Allwinner NAND Flash Controller.
		[jmcneill 20171113]
	gcc: Import GCC 5.5.  [mrg 20171113]
	libtre: Update to the latest git source. [rin 20171117]
	qemufwcfg(4): Add driver for QEMU Firmware Configuration device.
		[jmcneill 20171125]
	gdb(1): Updated to 8.0.1.  [christos 20171128]
	dhcpcd(8): Import dhcpcd-7.0.0-rc4. [roy 20171206]
	ihidev(4), ims(4): Added drivers for i2c HID mice. [bouyer 20171210]
	lua: Applied 6th bugfix to Lua 5.3.4 from lua.org/bugs.html.
		[mbalmer 20171213]
	sstk: Remove the "stack section size change" syscall [kamil 20171219]
	sbrk: Remove the "data section size change" syscall [kamil 20171219]
	ovadvise: Remove the old vadvise syscall header [kamil 20171219]
	dhcpcd(8): Import dhcpcd-7.0.0 [roy 20180101]
	mdnsd(8), dns-sd(1), libdns_sd: import mDNSResponder 878.1
		[christos 20180114]
	cpuctl(8): Add cpuctl identify support for sparc and sparc64.
		[mrg 20180116]
	macppc: Enable Veriexec support in the GENERIC kernel [sevan 20180117]
	raid(4): Add support for 32-bit on 64-bit kernel.  [mrg 20180117]
	zoneinfo: Import tzdata2018c (skipping 2018a & 2018b). [kre 20180124]
	libc: Update to tzcode2018c. [christos 20180125]
	i386: Add support for SMAP. [maxv 20180128]
	dhcpcd(8): Import dhcpcd-7.0.1 [roy 20180129]
	gcc: Import GCC 6.4.  [mrg 20180201]
	openldap: Import 2.4.45. [christos 20180205]
	unbound: Import 1.6.8. [christos 20180205]
	nsd: import 4.1.19 [christos 20180209]
	ddb(4):	Introduce dumpstack sysctl for printing a stack trace on panic,
		enable by default. [sevan 20180217]
	sdtemp(4): Add Microchp EMC1501, another device ID of Maxim MAX6604
		support. [msaitoh 20180222]
	ichsmb(4): Add support for Intel C620 devices. [msaitoh 20180222]
	macppc: Enable support for "per-priority cyclical scan" buffer queue
		strategy. [sevan 20180223]

lib/libc/hash/hmac.c
/*	$NetBSD: hmac.c,v 1.5 2017/10/05 09:59:04 roy Exp $	*/

/*-
 * Copyright (c) 2016 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Christos Zoulas.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <sys/cdefs.h>
__RCSID("$NetBSD: hmac.c,v 1.5 2017/10/05 09:59:04 roy Exp $");

#include <string.h>
#include <stdlib.h>

#include <md2.h>
#include <md4.h>
#include <md5.h>
#include <rmd160.h>
#include <sha1.h>
#include <sha2.h>

#define HMAC_SIZE	128
#define HMAC_IPAD	0x36
#define HMAC_OPAD	0x5C

static const struct hmac {
	const char *name;
	size_t ctxsize;
	size_t digsize;
	size_t blocksize;
	void (*init)(void *);
	void (*update)(void *, const uint8_t *, unsigned int);
	void (*final)(uint8_t *, void *);
} hmacs[] = {
	{
		"md2", sizeof(MD2_CTX), MD2_DIGEST_LENGTH, MD2_BLOCK_LENGTH,
		(void *)MD2Init, (void *)MD2Update, (void *)MD2Final,
	},
	{
		"md4", sizeof(MD4_CTX), MD4_DIGEST_LENGTH, MD4_BLOCK_LENGTH,
		(void *)MD4Init, (void *)MD4Update, (void *)MD4Final,
	},
	{
		"md5", sizeof(MD5_CTX), MD5_DIGEST_LENGTH, MD5_BLOCK_LENGTH,
		(void *)MD5Init, (void *)MD5Update, (void *)MD5Final,
	},
	{
		"rmd160", sizeof(RMD160_CTX), RMD160_DIGEST_LENGTH,
		RMD160_BLOCK_LENGTH,
		(void *)RMD160Init, (void *)RMD160Update, (void *)RMD160Final,
	},
	{
		"sha1", sizeof(SHA1_CTX), SHA1_DIGEST_LENGTH, SHA1_BLOCK_LENGTH,
		(void *)SHA1Init, (void *)SHA1Update, (void *)SHA1Final,
	},
	{
		"sha224", sizeof(SHA224_CTX), SHA224_DIGEST_LENGTH,
		SHA224_BLOCK_LENGTH,
		(void *)SHA224_Init, (void *)SHA224_Update,
		(void *)SHA224_Final,
	},
	{
		"sha256", sizeof(SHA256_CTX), SHA256_DIGEST_LENGTH,
		SHA256_BLOCK_LENGTH,
		(void *)SHA256_Init, (void *)SHA256_Update,
		(void *)SHA256_Final,
	},
	{
		"sha384", sizeof(SHA384_CTX), SHA384_DIGEST_LENGTH,
		SHA384_BLOCK_LENGTH,
		(void *)SHA384_Init, (void *)SHA384_Update,
		(void *)SHA384_Final,
	},
	{
		"sha512", sizeof(SHA512_CTX), SHA512_DIGEST_LENGTH,
		SHA512_BLOCK_LENGTH,
		(void *)SHA512_Init, (void *)SHA512_Update,
		(void *)SHA512_Final,
	},
};

static const struct hmac *
hmac_find(const char *name)
{
	for (size_t i = 0; i < __arraycount(hmacs); i++) {
		if (strcmp(hmacs[i].name, name) != 0)
			continue;
		return &hmacs[i];
	}
	return NULL;
}

ssize_t
hmac(const char *name,
    const void *key, size_t klen,
    const void *text, size_t tlen,
    void *digest, size_t dlen)
{
	uint8_t ipad[HMAC_SIZE], opad[HMAC_SIZE], d[HMAC_SIZE];
	const uint8_t *k = key;
	const struct hmac *h;
	uint64_t c[32];
	void *p;

	if ((h = hmac_find(name)) == NULL)
		return -1;


	if (klen > h->blocksize) {
		(*h->init)(c);
		(*h->update)(c, k, (unsigned int)klen);
		(*h->final)(d, c);
		k = (void *)d;
		klen = h->digsize;
	}

	/* Form input and output pads for the digests */
	for (size_t i = 0; i < sizeof(ipad); i++) {
		ipad[i] = (i < klen ? k[i] : 0) ^ HMAC_IPAD;
		opad[i] = (i < klen ? k[i] : 0) ^ HMAC_OPAD;
	}

	p = dlen >= h->digsize ? digest : d;
	if (p != digest) {
		memcpy(p, digest, dlen);
		memset((char *)p + dlen, 0, h->digsize - dlen);
	}
	(*h->init)(c);
	(*h->update)(c, ipad, (unsigned int)h->blocksize);
	(*h->update)(c, text, (unsigned int)tlen);
	(*h->final)(p, c);

	(*h->init)(c);
	(*h->update)(c, opad, (unsigned int)h->blocksize);
	(*h->update)(c, digest, (unsigned int)h->digsize);
	(*h->final)(p, c);

	if (p != digest)
		memcpy(digest, p, dlen);

	return (ssize_t)h->digsize;
}
